HarfangLab/harfanglab/iocs.py

```diff
@@ -0,0 +1,173 @@
+
+    def api_search_indicator(self, indicator_value: str):
+        response = self.client.get(
+            f"{self.client.instance_url}/api/data/threat_intelligence/IOCRule/",
+            params={"search": indicator_value},
+            timeout=60,
+        )
+        self._handle_response_error(response)
+        return response
+
+    def api_create_indicator(self, indicator: dict[str, Any]):
+        response = self.client.post(
+            f"{self.client.instance_url}/api/data/threat_intelligence/IOCRule/", json=indicator, timeout=60
+        )
+        self._handle_response_error(response)
+        return response
+
+    def get_reference_url(self, id) -> str:
+        return urljoin(self.sekoia_base_url, f"intelligence/objects/{id}")
+
+    @staticmethod
+    def get_payload(value: Any, type: Any, args: dict[str, Any]):
+        return {
+            "source_id": args["source_id"],
+            "type": type,
+            "value": value,
+            "block_on_agent": args["block_on_agent"],
+            "quarantine_on_agent": args["quarantine_on_agent"],
+            "endpoint_detection": args["detect_on_agent"],
+            "enabled": True,
+        }
+
+    def get_valid_indicators(self, stix_objects: Any, args: dict[str, Any]) -> dict[Any, Any]:
+        """
+        Transform the IOC from the STIX format to the payload expected by HarfangLab
+        """
+
+        seen_values: defaultdict[str, list[str]] = defaultdict(list)
+        results: dict[str, Any] = {"valid": [], "revoked": [], "expired": []}
+
+        for object in stix_objects:
+            # Extract value and type from pattern
+            self.log(message=f"object in stix_objects {str(object)}", level="debug")
+            indicators = stix_to_indicators(stix_object=object, supported_types_map=self.SUPPORTED_TYPES_MAP)
+            for indicator in indicators:
+                ioc_value = indicator["value"]
+                ioc_type = indicator["type"]
+                result = self.get_payload(value=ioc_value, type=ioc_type, args=args)
+
+                # Handle revoked objects
+                if object.get("revoked", False):
+                    results["revoked"].append(result)
+                    continue
+
+                # Ignore this IOC if the same value has already been processed
+                if ioc_value in seen_values[ioc_type]:
+                    continue
+
+                # Add a direct link in description if the data is originating from Sekoia.io
+                if "x_ic_observable_types" in object.keys():
+                    result["description"] = self.get_reference_url(object["id"])
+
+                # Compare expiration data if exists
+                valid_until = object.get("valid_until")
+                if valid_until:
+                    current_datetime = datetime.now(timezone.utc)
+                    valid_until_datetime = isoparse(valid_until)
+                    if valid_until_datetime >= current_datetime:
```

Comment:
The comparison is inverted:
```suggestion
                    if valid_until_datetime < current_datetime:
```

Reply:
😅 you're right, thanks